location/motif.py

```diff
@@ -833,6 +833,129 @@ def compute_nodes(df,
     return df, nodes
 
 
+def generate_motifs(data,
+                    nodes,
+                    sr_col='stay_region',
+                    round_trip=True,
+                    insert_home=True,
+                    home=None,
+                    dayofweek=[0, 1, 2, 3, 4],
+                    trav_dist_th=50000,
+                    valid_timeslot_th=8):
+    """
+    Generate moitfs for given data.
+
+    Parameters:
+    -----------
+
+    data: DataFrame
+        Location data.
+
+    nodes: tuple
+        Nodes generated by generate_daily_nodes().
+
+    sr_col: str
+        Column name for stay region.
+        Default is 'stay_region'.
+
+    round_trip: bool
+        Whether to consider only days that start and end at the same date.
+        Default is True.
+
+    insert_home: bool
+        Whether to insert home location if the first timeslot of daily
+        data is missing. Default is True.
+
+    home: str
+        Home location in geohash form.
+        Default is None. In this case, home locatoin is approximated using
+        user location data.
+
+    dayofweek: list of integers
+        Which days in a week to consider, [0,1,2,3,4,5,6] for Mon, Tue, Wed,
+        Thurs, Fri, Sat, and Sun. Defualt is weekdays.
+
+    trav_dist_th: int
+        Travel distance threshold used to filter out days on which the user
+        travels to other cities. Default is 50,000 meters (about 31 miles).
+
+    valid_timeslot_th: int
+        Valid time slot thresold required to compute daily motifs.
+        Defualt is 8 intervals.
+
+
+    Returns:
+    --------
+
+    motifs: list of dictionary
+        List of motifs, key is a graph object, value is the list of timestamp
+        for days having the same motif
+    """
+
+    loc_data = data.copy()
+
+    # compute home location if needed
+    if home is None:
+        home = utils.approx_home_location(loc_data, sr_col=sr_col)
+
+    # generate motifs
+    motifs = []
+
+    for n in nodes:
+        tsp = n[0]  # timestamp for current daily nodes
+        list_nodes = n[1].node.dropna().tolist()
+
+        # check day of week
+        if n[0].dayofweek not in dayofweek:
+            continue
+
+        # check number of valid time slots
+        if len(list_nodes) <= valid_timeslot_th:
+            continue
+
+        # Fiter out days with abnormal travel distance
+        # Skip current daily nodes if the maximum of pair distances among daily
+        # nodes exceed the threshold.
+        # That is, disgard the data if the user's trajectory is beyond
+        # the area of one city
+
+        different_visited_locations = np.unique(list_nodes)
+        dist_list = [vincenty(geohash.decode(x), geohash.decode(home)).m
+                     for x in different_visited_locations]
+        if any(d > trav_dist_th for d in dist_list):
+            continue
+
+        # insert home location is necessary
```

Comment:
Not sure what is happening here. If `insert_home` is true, you should just insert home, right? Not sure about the extra condition